### PR TITLE
Grafana/Live: Adds function for retrieving promise-bound live srv

### PR DESCRIPTION
**Pull request by kaydelaney**
**What this PR does / why we need it**:
Adds a function, `getConnectedLiveSrv`, that returns a promise which resolves with the singleton live srv instance once it has connected. This should help simplify usage of the live srv in certain instances (as with LivePanel.tsx)

### CODE REVIEW

File: packages/grafana-runtime/src/services/live.ts [hunk 1]
```diff
@@ -48,6 +48,8 @@ export interface GrafanaLiveSrv {
    * Send data to a channel.  This feature is disabled for most channels and will return an error
    */
   publish<T>(channel: string, data: any): Promise<T>;
+
+  addListener(event: string | symbol, listener: (...args: any[]) => void): void;
```
Comment (ryantxu) [inline packages/grafana-runtime/src/services/live.ts:52]: what are possible event values?  for our API, should they be more specific than string|symbol?

File: packages/grafana-runtime/src/services/live.ts [hunk 2]
```diff
@@ -70,3 +72,18 @@ export const setGrafanaLiveSrv = (instance: GrafanaLiveSrv) => {
...
+/**
+ * Returns a promise that resolves with the {@link GrafanaLiveSrv} once it has
+ * connected
+ *
+ * @experimental
+ * @public
+ */
+export const getConnectedLiveSrv = (): Promise<GrafanaLiveSrv> => {
+  return new Promise(resolve => {
+    singletonInstance.addListener('connect', () => {
```
Comment (ryantxu) [inline packages/grafana-runtime/src/services/live.ts:85]: what happens if it is already connected?

Comment (ryantxu) [reply packages/grafana-runtime/src/services/live.ts:85]: should we remove the listener after it is resolved?

File: packages/grafana-runtime/src/services/live.ts (hunk 1 above)
Comment (kaydelaney) [reply packages/grafana-runtime/src/services/live.ts:52]: This is a good question, I just copied the call signature of the corresponding centrifuge method but you're right that there could be more specificity. I'll look into it

File: packages/grafana-runtime/src/services/live.ts (hunk 2 above)
Comment (kaydelaney) [reply packages/grafana-runtime/src/services/live.ts:85]: Another good catch. Should add a check to see if we're already connected

File: packages/grafana-runtime/src/services/live.ts (hunk 1 above)
Comment (kaydelaney) [reply packages/grafana-runtime/src/services/live.ts:52]: It seems like arbitrarily named events can be emitted with `centrifuge.emit`, but I suppose from the perspective of a consumer of the grafana live API that flexibility might not be needed or advisable. We could limit it just to the events that centrifuge-js itself makes use of (publish, join, leave, subscribe, unsubscribe, error, message, connect, disconnect)?

Comment (torkelo) [reply packages/grafana-runtime/src/services/live.ts:52]: Not sure what the purpose of this listener is we have the getChannelStream for this exact purpose no?

File: packages/grafana-runtime/src/services/live.ts (hunk 2 above)
Comment (kaydelaney) [reply packages/grafana-runtime/src/services/live.ts]: I like this design! Pushed a change

File: packages/grafana-runtime/src/services/live.ts [hunk 3]
```diff
@@ -70,3 +75,15 @@ export const setGrafanaLiveSrv = (instance: GrafanaLiveSrv) => {
...
+ * Returns a promise that resolves with the {@link GrafanaLiveSrv} once it has
+ * connected
+ *
+ * @experimental
+ * @public
+ */
+export const getConnectedLiveSrv = async () => {
+  await singletonInstance.waitUntilConnected();
+  return singletonInstance;
+};
```
Comment (torkelo) [inline packages/grafana-runtime/src/services/live.ts:89]: Looking at how we use the currently it does not look like we need two getLiveSrv functions. The connect happens in the constructor. So everyone who is interested in the LiveSrv should probably use getConnectedLiveSrv, so I would make this the default behavior. and only have a single getGrafanaLiveSrv 

But I also question a bit the need for this in the first place. Can't we make initChannel / getChannelStream / publish handle this connection wait internally?

Comment (torkelo) [reply packages/grafana-runtime/src/services/live.ts:89]: but having this as default could be a start

Comment (kaydelaney) [reply packages/grafana-runtime/src/services/live.ts:89]: This is fair actually, let me make some changes

File: public/app/features/live/live.ts [hunk 4]
```diff
@@ -114,9 +115,19 @@ class CentrifugeSrv implements GrafanaLiveSrv {
   }
 
   initChannel<T>(path: string, handler: ChannelHandler<T>) {
+    const connected = from(
+      new Promise<void>(resolve => {
+        if (this.centrifuge.isConnected()) {
+          return resolve();
+        }
+
+        this.centrifuge.addListener('connect', resolve);
```
Comment (torkelo) [inline public/app/features/live/live.ts:124]: you should be able to define this promise in the constructor and reuse it

Comment (torkelo) [reply public/app/features/live/live.ts:124]: as a promise that has completed can be reused  (and will complete with same value)

**Review by torkelo — APPROVED** [review]
Looks good, but would move the connected promise to an instance property so it can be reused (like a cache of is connected state).

File: public/app/features/live/live.ts [hunk 5]
```diff
@@ -133,15 +144,17 @@ class CentrifugeSrv implements GrafanaLiveSrv {
       },
     };
     c.subscription = this.centrifuge.subscribe(path, callbacks);
+    return connected;
   }
 
   getChannelStream<T>(path: string): Observable<T> {
```
Comment (ryantxu) [inline public/app/features/live/live.ts:150]: The original idea is that a few things may connect to the same stream -- in an alternate PR, I tried:

https://github.com/grafana/grafana/pull/27583/files#diff-234f06c46c08d978307cf6c087712a55R134

Comment (torkelo) [reply public/app/features/live/live.ts:150]: that should still be possible I think

File: packages/grafana-runtime/src/services/live.ts (hunk 3 above)
Comment (hugohaggmark) [reply packages/grafana-runtime/src/services/live.ts]: I agree with Torkel, as a consumer I'm only interested in getting a stream to subscribe/publish too.

**Review by torkelo — APPROVED** [review]

**Pull request closed without merging**